pyseir/cli.py

```diff
@@ -74,21 +62,97 @@ def _states_region_list(state: Optional[str], default: List[str]) -> List[pipeli
-    region: pipeline.Region, run_mode=DEFAULT_RUN_MODE, output_interval_days=1, output_dir=None,
-) -> model_fitter.ModelFitter:
-    assert region.is_state()
+@dataclass
+class StatePipeline:
+    """Runs the pipeline for one state and stores the output."""
+
+    region: pipeline.Region
+    infer_df: pd.DataFrame
+    fitter: model_fitter.ModelFitter
+
+    @staticmethod
+    def run(region: pipeline.Region, run_mode: pyseir.utils.RunMode) -> "StatePipeline":
+        assert region.is_state()
+        infer_df = infer_rt.run_rt(infer_rt.RegionalInput.from_region(region))
+        fitter_input = model_fitter.RegionalInput.from_state_region(region)
+        fitter = model_fitter.ModelFitter.run_for_region(fitter_input)
+        ensembles_input = ensemble_runner.RegionalInput.for_state(fitter)
+        ensemble_runner.run_region(ensembles_input, ensemble_kwargs={"run_mode": run_mode})
+        return StatePipeline(region=region, infer_df=infer_df, fitter=fitter)
+
+
+@dataclass
+class SubStateRegionPipelineInput:
+    region: pipeline.Region
+    run_fitter: bool
+    state_fitter: model_fitter.ModelFitter
+    run_mode: pyseir.utils.RunMode
+
+    @staticmethod
+    def build_all(
+        fips: Optional[str],
+        run_mode: pyseir.utils.RunMode,
+        state_fitter_map: Mapping[pipeline.Region, model_fitter.ModelFitter],
+    ) -> List["SubStateRegionPipelineInput"]:
+        """For each region smaller than a state, build the input object used to run the pipeline."""
+        # TODO(tom): Pass in the combined dataset instead of reading it from a global location.
+        # Calculate the whitelist for the infection rate metric which makes no promises
+        # about it's relationship to the SEIR subset
+        if fips:  # A single Fips string was passed as a flag. Just run for that fips.
+            infer_rt_regions = {pipeline.Region.from_fips(fips)}
+        else:  # Default to the full infection rate whitelist
+            infer_rt_regions = {
+                pipeline.Region.from_fips(x)
+                for x in combined_datasets.load_us_latest_dataset().all_fips
+                if len(x) == 5
+                and "25" != x[:2]  # Counties only  # Masking MA Counties (2020-08-27) due to NaNs
+                and "999" != x[-3:]  # Remove placeholder fips that have no data
+            }
+        # Now calculate the pyseir dependent whitelist
+        whitelist_df = _generate_whitelist()
+        all_county_regions = set(
+            whitelist.regions_in_states(
+                list(state_fitter_map.keys()), fips=fips, whitelist_df=whitelist_df
+            )
+        )
+        pipeline_inputs = [
+            SubStateRegionPipelineInput(
+                region=region,
+                run_fitter=(region in all_county_regions),
+                state_fitter=state_fitter_map.get(region.get_state_region()),
+                run_mode=run_mode,
+            )
+            for region in (infer_rt_regions | all_county_regions)
```

Comment:
I don't think the union of these sets is the logic we want here. For example, the pipeline should route the regions representing MA counties to the SEIR_model module, but not to the infection_rate module.

Reply:
There was a bug in my PR when you looked at it: `SubStatePipeline.run` wasn't respecting `SubStateRegionPipelineInput.run_fitter`. It is also possible the code that sets run_fitter (just above where you added this comment) isn't clear.